### DISCUSSION TITLE
Successful install on Debian 12 (Bookworm) i.e. how to work with/around PEP-668.

### DISCUSSION THREAD

**Discussion by jaimet** [qtile/qtile#4542] [Show and tell]
Hi all. Complete qtile newbie here.

After several late evenings and much gnashing of teeth, I have finally succeeded in installing qtile (0.23.0) on debian 12 (bookworm). I am documenting what I've learnt because I wish that someone else had already documented this for me to read before I started trying to get this to work.

### Background (no more `pip install ...`)
Despite [a brief near-miss](https://tracker.debian.org/pkg/qtile) and a semi-current ["Intent to package"](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1015267), qtile has never been included in the official Debian distribution so debian's _usual_ package installation method (`sudo apt install qtile`) will not work. Instead, qtile's [recommended installation method ](https://docs.qtile.org/en/v0.23.0/manual/install/ubuntu.html) has, up until now, been `pip install ...`.

However, [Debian](https://www.debian.org/releases/bookworm/amd64/release-notes/ch-information.en.html#python3-pep-668), [Ubuntu](https://pythonspeed.com/articles/externally-managed-environment-pep-668/) and [Fedora](https://fedoraproject.org/wiki/Changes/PythonMarkExternallyManaged) have all recently made stable releases which incorporate [PEP-668](https://peps.python.org/pep-0668/) - this is a change to the python ecosystem designed to avoid [system environment breakage caused by using potentially-conflicting package managers](https://pythonspeed.com/articles/externally-managed-environment-pep-668/) (for example `sudo apt install` vs. `pip install`). One consequence of this change is that we can no longer use `pip install ...` to install system packages as we did previously.

### Method 1: `break-system-packages` (no thanks!)

[PEP-668](https://peps.python.org/pep-0668/) is basically a safety mechanism that stops pip from installing into your python "system environment". There are several different ways to override (i.e. _ignore_) this safety mechanism:

1) Run `pip` with the `--break-system-packages option`, or
2) Run `pip` with the environment variable `PIP_BREAK_SYSTEM_PACKAGES` set to `1`, or
3) Add the line `break-system-packages = true` to your pip config (`~/.config/pip/pip.conf` or `/etc/pip.conf`), or
4) Delete `/usr/lib/python3.*/EXTERNALLY-MANAGED`.

While this approach is chosen by many[^1][^2][^3], I want to avoid it _if possible_. I personally think that PEP-668 is a move in the right direction, and that letting `pip` modify your system environment is a retrograde step (When your system breaks, you _will_ get to keep both pieces).

### Method 2: `venv` virtual environment plus `pip` (this failed)

(Important: this section is only here to document what **did not work**. If you want a working solution, see "Method 3" below.)

My first attempt used a `venv` virtual environment plus `pip`. After a minimal[^4] installation of debian, I installed qtile:
```
mkdir -p ~/.local/bin/qtile && cd $_
su root -c 'apt install python3-venv'
# (If you prefer, replace the above with sudo apt ...)
python3 -m venv venv
. venv/bin/activate
pip install qtile
deactivate
```
Then I installed the required xorg packages...
```
su root -c 'apt install --no-install-recommends \
  xserver-xorg-core xserver-xorg-input-libinput xinit mate-terminal'
# (Again, if you prefer, replace the above with sudo apt ...)
# Note to self: mate-terminal pulls in the necessary cairo and pango packages)
```
... and created `.xinitrc`:
```
cat << 'EOF' > ~/.xinitrc
. ~/.local/bin/qtile/venv/bin/activate
exec qtile start
EOF
# Note to self: shebang unnecessary, executable mode bit unnecessary
```
When I ran `startx`, I got the default qtile desktop, but there was a problem: `which python3` (run from inside a terminal emulator) returned the virtual environment's python interpreter:
```
user@laptop:~$ which python3
/home/user/.local/bin/qtile/venv/bin/python3
```
Hmmm. Could I `deactivate` this partially-inherited virtual environment? No:
```
user@laptop:~$ deactivate
bash: deactivate: command not found
```
Not good. It appears that terminal emulators (spawned by qtile) inherit _parts_ of qtile's underlying virtual environment (such as the interpreter path), but they do not inherit enough of that environment to be able to `deactivate` it.[^5] I decided to try a different approach...
### Method 3: `pipx` (this succeeded!)

Next, I tried [Debian 12's release notes' suggestion](https://www.debian.org/releases/bookworm/amd64/release-notes/ch-information.en.html#python3-pep-668): using `pipx`.

After a minimal[^4] installation of debian, install qtile:
```
su root -c 'apt install --no-install-recommends pipx'
# (If you prefer, replace the above with sudo apt ...)
pipx install qtile
```
(If the `pipx install qtile` line produces an error message saying that a necessary directory `is not on your PATH environment variable` then simply log-out and log back in again - debian 12 already includes the required code in `.profile`.)

Then install the required xorg packages...
```
su root -c 'apt install --no-install-recommends \
  xserver-xorg-core xserver-xorg-input-libinput xinit mate-terminal'
# (Again, if you prefer, replace the above with sudo apt ...)
# Note to self: mate-terminal pulls in the necessary cairo and pango packages)
```
... and create the "pipx version" of `.xinitrc`:
```
cat << 'EOF' > ~/.xinitrc
exec qtile start
EOF
# Note to self: shebang unnecessary, executable mode bit unnecessary
```
Now, when I run `startx`, I get the default qtile desktop, and `which python3` (run from inside a terminal emulator) returns the "correct" python interpreter:
```
user@laptop:~$ which python3
/usr/bin/python3
```
HTH, jaimet

[^1]: [JustAGuy Linux's Speed Challenge](https://www.youtube.com/watch?v=GUfpYMih4s8) includes qtile - [his installation script](https://github.com/drewgrif/bookworm-scripts/blob/main/resources/qtile-commands) does a `sudo rm /usr/lib/python3.11/EXTERNALLY-MANAGED`.
[^2]: [JerrySM64's qtile installation script](https://github.com/JerrySM64/Qtile-Debian/blob/main/qtile-setup.sh) also does a `sudo rm /usr/lib/python3.11/EXTERNALLY-MANAGED`. 
[^3]: https://www.reddit.com/r/debian/comments/14g9np9
[^4]: At the "Software selection" (tasksel) screen, I select only "SSH server" and "standard system utilities".
[^5]: I don't think that this is a problem with `qtile` in isolation, as the `pipx` approach succeeds. It may be a "`qtile` in a `venv`" _architecture_ problem, but more research is needed.

**Comment by m-col** [comment]
Excellent write up, thanks!

**Comment by alfredojoseneto** [comment]
Someone had problem with Pyenv?

**Comment by jaimet** [comment]
> Someone had problem with Pyenv?

I have never tried/studied `pyenv` but, from what I have read in the last 10 minutes, I am not sure whether this would help with Debian's "adoption"/"incorporation" of PEP-668.

Are you suggesting `pyenv` as a way to help install `qtile` on Debian 12? If yes, have you succeeded?

**Comment by alfredojoseneto** [comment]
I tried to install and something didn't work well. The `tile` was not recognize the correct Python and I don't know if `pyenv` is the responsible for this or myself (hahahaha, probably the second choice).
But I'll try the suggestion with `pipx` and see what happens. Thanks in advance

**Comment by asoymos** [comment]
Thank you! made my life easier. Would like to complement with a little tip. Qtile was complaining because dbus-next wasn't installed, I used `apt` but it wasn't recognized. Solution was to install inside venv for Qtile like this:
`pipx inject qtile dbus-next`
You can also install other modules for widgets like psutils in same way. Enjoy!
